evaltools/data/estimatecvap.py

```diff
@@ -0,0 +1,258 @@
+        if not (cvap10 in acs_source or cvap10 in cvap_source):
+            possible_columns = set(acs_source).union(set(cvap_source))
+            raise ValueError(
+                f"Your CVAP column '{cvap10}' must be contained in either the ACS "
+                f"or Special Tab columns: {possible_columns}"
+            )
+        
+        if not vap10 in acs_source:
+            raise ValueError(
+                f"Your old VAP column '{vap10}' must be contained in the ACS "
+                "columns: {set(acs_source)}"
+            )
+
+        # If the VAP20 column passed doesn't exist on the user-provided geometries,
+        # raise an error.
+        if not vap20 in base:
+            raise ValueError(
+                f"Your new VAP column '{vap20}' must be contained in your base " +
+                f"dataframe: {set(base)}"
+            )
+    
+    # Remove ACS 5 columns that overlap with special-tab ones.
+    non_overlaps = list(set(acs_source).difference(set(cvap_source)))
+    acs_source = acs_source[[cvap_geoid] + non_overlaps]
+    source = cvap_source.merge(acs_source, on=cvap_geoid)
+
+    # Get the right columns from the base geometry, and map the base geometries
+    # to the units with CVAP data. Apparently dropping bad columns by using slicing
+    # incudes a SettingWithCopy warning, so we're just dropping using .drop()
+    # instead.
+    correct = ["geometry"] + [col for col in list(base) if any(sub in col for sub in ["POP", "VAP", "GEOID"])]
+    bads = list(set(base) - set(correct))
+    pared = base.drop(bads, axis=1)
+    pared = mapbase(pared, state, geometry10)
+
+    # Compute weights.
+    for (cvap10, vap10, _) in groups: 
+       source[cvap10 + "%"] = source[cvap10]/source[vap10]
+
+    # Fill in values according to the following rules:
+    # 
+    #   1.  if there are 0 *CVAP reported and 0 *VAP reported, we set the weight to
+    #       the average *CVAP/*VAP ratio within the county;
+    #   2.  if there are 0 *CVAP reported and *VAP > 0, we set the weight to zero_fill;
+    #   3.  if *CVAP > 0 but *VAP = 0 or *CVAP/*VAP > percentage_cap, we set the weight to 1.
+    statewide = {
+        cvap + "%": source[cvap].sum()/source[vap].sum() if source[vap].sum() != 0 else 0
+        for (cvap, vap, _) in groups
+    }
+
+    # Rename colunms with percentages.
+    cvappcts = [cvap + "%" for (cvap, _, _) in groups]
+
+    # Get the county names and compute population-weighted CVAP averages. This
+    # serves as a replacement for the statewide average.
+    source["_county"] = source[cvap_geoid].str[2:5]
+    counties = list(set(source["_county"]))
+    countyaverages = { pct: {} for pct in cvappcts }
+
+    for county in counties:
+        chunk = source[source["_county"] == county]
+        
+        for cvap19, vap19, _ in groups:
+            cvap19total = chunk[cvap19].sum()
+            vap19total = chunk[vap19].sum()
+            countyaverages[cvap19 + "%"][county] = cvap19total/vap19total
+
```

Comment:
Here we compute the county averages for each CVAP column only once, and can then apply them based on the county identifier of each `base` unit.